### PR TITLE
Add datascript as a data cache

### PR DESCRIPTION
**Pull request by acron0**
**_But, why?**_

I think keeping a single 'app state with everything in it' is a naive approach to handling React applications. At any one time my prediction is that we'll have more data than we need to physically display (prime example being all a user's available projections). Datascript sits as a cache between 'data fetched from the server' and 'what the frontend is currently interested in' and obviously makes things more manageable at the app state level without the latency trade-off of having to hit the server every time we need to fetch data that we've had to jettison.

### CODE REVIEW

File: src/witan/schema/core.cljc [hunk 1]
```diff
@@ -19,11 +20,15 @@
...
+   (s/required-key :type)          ProjectionTypes
+   (s/required-key :owner)         s/Str
+   (s/required-key :version)       s/Int
+   (s/required-key :last-modified) s/Str
+   (s/required-key :last-modifier) s/Str
+   (s/optional-key :descendant-id) ProjectionIdType
+   ;; added internally
+   (s/optional-key :db/id)         s/Int
+   (s/optional-key :has-ancestors) s/Bool
+   })
```
Comment (otfrom) [inline src/witan/schema/core.cljc]: tidy up trailing parens. Generally probably worth running cljfmt over given as we are starting out.

File: src/witan/styles/base.clj [hunk 2]
```diff
@@ -148,5 +148,8 @@
 
          [:.witan-projection-table-row-selected
           :.witan-projection-table-row-selected:hover
-          {:background-color colour/row-selected}]])
+          {:background-color colour/row-selected}]
+
+
+         ])
```
Comment (otfrom) [inline src/witan/styles/base.clj]: whitespace tidy

File: src/witan/ui/components/dashboard.cljs [hunk 3]
```diff
@@ -31,17 +31,17 @@
                ;;
                (if (not (empty? selected))
```
Comment (otfrom) [inline src/witan/ui/components/dashboard.cljs]: How about not-empty ?

https://clojuredocs.org/clojure.core/not-empty

File: src/witan/ui/components/dashboard.cljs [hunk 4]
```diff
@@ -31,17 +31,17 @@
...
                  [:li.witan-menu-item.pure-menu-item
-                  [:a {:href (str "#/projection/" (:id selected))}
+                  [:a {:href (str "#/projection/" (second selected))}
```
Comment (otfrom) [inline src/witan/ui/components/dashboard.cljs]: worth destructuring or pulling out in the let above?

File: src/witan/ui/controllers/input.cljs [hunk 5]
```diff
@@ -10,5 +11,17 @@
...
+
+(defmethod handler
+  :event/toggle-tree-view
+  [[event args] cursor]
+  (s/validate Projection args)
+  (let [db-id        (-> args :db/id)
+        id           (-> args :id)
+        is-toggled   (contains? (-> cursor :projections-meta :expanded) [db-id id])
+        fn           (if is-toggled disj conj)
+        new-state    (om/transact! cursor [:projections-meta :expanded] #(fn % [db-id id]))
```
Comment (otfrom) [inline src/witan/ui/controllers/input.cljs]: Ooh, I like that fn in the if and then the disj conj usage

File: src/witan/ui/data.cljs [hunk 6]
```diff
@@ -1,9 +1,103 @@
...
+  (let [top-level (apply concat (d/q '[:find (pull ?e [*])
+                                       :where [?e :id _]
+                                       [(get-else $ ?e :descendant-id nil) ?u]
+                                       [(nil? ?u)]]
+                                     @db-conn))]
+    (if-not expand
+      top-level
+      (if (vector? expand)
+        (mapcat (fn [projection]
+                  (if (first (filter (fn [[ck cv]] (= (ck projection) cv)) expand))
```
Comment (otfrom) [inline src/witan/ui/data.cljs]: Use some ?

https://clojuredocs.org/clojure.core/some#example-542692c6c026201cdc326940

File: src/witan/ui/controllers/input.cljs (hunk 5 above)
Comment (acron0) [inline src/witan/ui/controllers/input.cljs]: The beauty of first-class functions.

**Comment by otfrom** [comment]
:shipit:

**Pull request merged**